src/client-side-metrics/operation-metrics-collector.ts

```diff
@@ -0,0 +1,337 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as fs from 'fs';
+import {IMetricsHandler} from './metrics-handler';
+import {MethodName, StreamingState} from './client-side-metrics-attributes';
+import {grpc} from 'google-gax';
+
+/**
+ * An interface representing a Date-like object.  Provides a `getTime` method
+ * for retrieving the time value in milliseconds.  Used for abstracting time
+ * in tests.
+ */
+interface DateLike {
+  /**
+   * Returns the time value in milliseconds.
+   * @returns The time value in milliseconds.
+   */
+  getTime(): number;
+}
+
+/**
+ * An interface representing a tabular API surface, such as a Bigtable table.
+ */
+export interface ITabularApiSurface {
+  instance: {
+    id: string;
+  };
+  id: string;
+  bigtable: {
+    appProfileId?: string;
+    clientUid: string;
+  };
+}
+
+/**
+ * Information about the completion of a single attempt of a Bigtable operation.
+ * This information is used for recording metrics.
+ */
+interface OnAttemptCompleteInfo {
+  connectivityErrorCount: number;
+  /**
+   * Whether the operation is a streaming operation or not.
+   */
+  streamingOperation: StreamingState;
```

Comment:
Why is this part of OnAttemptComplete? Isn't it known in advance?

My understanding is this is basically just to check if the user called the public `read_row` vs `read_rows` method, which should be known at call time? I'm not sure if that even applies to node though, since I'm not seeing a point read method.

How were you planning on populating this?

Reply:
>My understanding is the streaming flag is basically just to check if the user called the public read_row vs read_rows method, which should be known at call time?

That's my understanding too. This is now populated when the metrics collector gets created.